File: Daily/sbgo/Leetcode1449-数位成本为目标和的最大值.go
package sbgo

import "math"

func largestNumber(cost []int, target int) string {
	dp := make([]int, target+1)
	for i := 1; i <= target; i++ {
		dp[i] = math.MinInt32
	}
	for i := 0; i < 9; i++ {
		for j := cost[i]; j <= target; j++ {
			dp[j] = max(dp[j], dp[j-cost[i]]+1)
		}
	}

	if dp[target] < 0 {
		return "0"
	}
	ans := []byte{}
	j := target
	for i := 8; i >= 0; i-- {
		for j >= cost[i] && dp[j] == dp[j-cost[i]]+1 {
			ans = append(ans, byte('1'+i))
			j -= cost[i]
		}
	}
	return string(ans)
}

File: Daily/PY/Leetcode1449-数位成本为目标和的最大值.py
from typing import List

class Solution:
    def largestNumber(self, cost: List[int], target: int) -> str:
        MINI = int(-1e9 + 10)
        dp = [MINI] *(target + 1)
        dp[0] = 0
        for i in range(9):
            for j in range(cost[i],target + 1):
                dp[j] = max(dp[j],dp[j - cost[i]] + 1)

        if dp[target] < 0:
            return '0'
        ans = list()
        j = target
        for i in range(8,-1,-1):
            while j >= cost[i] and dp[j] == dp[j - cost[i]] + 1:
                ans.append(i + 1)
                j -= cost[i]
        return ''.join(ans) 